### PR TITLE
Update the maneuver icons for notifications

### LINKED ISSUE #5442: Update the maneuver icons for notifications
As a result of this [change](https://github.com/mapbox/mapbox-navigation-android/pull/5430), the maneuver icons in the module `libnavui-maneuver` is now different than the icons in `libtrip-notification`. This ticket tracks changing the icons for notification.

Some more [history](https://github.com/mapbox/mapbox-navigation-android/pull/4683#issuecomment-892245936) as to why the modules maintain same set of icons in 2 different places.

cc @tomaszrybakiewicz @Guardiola31337

### PR DESCRIPTION
**Pull request by tomaszrybakiewicz**
Closes #5442 

### Description

**TL;DR** 
This PR does 3 things:
1. Adds new maneuver icons for notifications.
2. Fixes _libnavui-maneuver_ to correctly "driving side flip" recently added turn icons.
3. Removes duplicate `NotificationTurnIconHelper` by moving up _libnavui-maneuver_ `TurnIconHelper` to shared _libnavui-base_ module.

| Current TurnIconHelper setup | New TurnIconHelper setup |
|:---:|:---:|
|TurnIconHelper-prev|TurnIconHelper|

**Detailed module changes**
_libnavui-base:_
- Added `TurnIconResource` interface defining all icon resources used by both _libnavui-maneuver_ and _libtrip-notification_ modules.
- Moved to internal package `TurnIconHelper` class from the `libnavui-maneuver` 
   - Refactored `TurnIconHelper` class  to use single source of truth for both turn and roundabout icons.
   - Simplified `ManeuverTurnIcon` create logic
   - Added driving side support for new icons.
 
_libnavui-maneuver:_
- Updated  the `TurnIconResource` class to be an implementation of the  _libnavui-base_ module `TurnIconResource` interface.
   - Added static `defaultIconSet()` factory method.
- Removed the `TripIconHelper` module class in favor of  _libnavui-base_ module `TripIconHelper` class.
- Updated `MapboxTurnIconsApi` class to use _libnavui-base_ module `TripIconHelper` class.

_libtrip-notification_
- Renamed `NotificationTurnIconResource` class to `TripIconResource`
   - Updated this class be an implementation of _libnavui-base_ module `TurnIconResource` interface.
   - Added static `defaultIconSet()` factory method.
- Removed the `TripIconHelper` module class in favor of  _libnavui-base_ module `TripIconHelper` class.
- Updated `MapboxTripNotification` class to use _libnavui-base_ module `TripIconHelper` class.

## Changelog
```
<changelog>Updated maneuver icons for notifications.</changelog>
```

### CODE REVIEW

**Review by Guardiola31337 — COMMENTED** [review]
Left a minor NIT.

File: libnavui-maneuver/src/test/java/com/mapbox/navigation/ui/maneuver/TurnIconHelperTest.kt [hunk 1]
```diff
@@ -1335,4 +1338,54 @@ class TurnIconHelperTest {
...
         assertEquals(expected, actual)
     }
+
+    @Test
+    fun `generate roundabout icon`() {
+        val actual = turnIconHelper.retrieveTurnIcon(ROUNDABOUT, 210f, SLIGHT_RIGHT, "right")
+
+        val expected =
+            ManeuverTurnIcon(210f, "right", false, R.drawable.mapbox_ic_roundabout_slight_left)
+        assertEquals(expected, actual)
```
[the remark this answers is not in the document]
Comment (tomaszrybakiewicz) [reply libnavui-maneuver/src/test/java/com/mapbox/navigation/ui/maneuver/TurnIconHelperTest.kt:1348]: NIT? I am confused. What should I change here?

**Comment by abhishek1508** [comment]
We should probably move the files you put under `libnavui-base` to `libnavigation-base`. The reason being if we look at the `build.gradle` for `libnavui-base`, we will see that it depends on `libnavigation-base` and `libnavigation-core`. Further if we add `libnavui-base` as a dependency to `libtrip-notification`, we are adding `libnavigation-core` as a dependency to `libtrip-notification` transitively. In general, if we are making any changes to files that affects components in core, it should be hosted in `libnavigation-base`.

/cc @tomaszrybakiewicz

**Comment by abhishek1508** [comment]
I think we might have overlooked an implementation details in v2. Looking at `MapboxTripNotification`

```
@MapboxModule(MapboxModuleType.NavigationTripNotification)
class MapboxTripNotification constructor(
    private val navigationOptions: NavigationOptions,
    private val distanceFormatter: DistanceFormatter
) : TripNotification {
....
....
private val turnIconHelper = TurnIconHelper(TurnIconResources.defaultIconSet())
}
```
We can see that we are not allowing external users to define their own set of `TurnIconResources` as we are defining it explicitly inside the class thereby losing the purpose of `TurnIconResources`. I think we should take this as a constructor parameter with default value being `TurnIconResources.defaultIconSet()` so it doesn't breaks SEMVER.

/cc @tomaszrybakiewicz @LukasPaczos

File: libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/internal/TurnIconResources.kt [hunk 2]
```diff
@@ -2,96 +2,105 @@ package com.mapbox.navigation.trip.notification.internal
...
-    @DrawableRes val turnIconTurnStraight: Int,
-    @DrawableRes val turnIconTurnSlightLeft: Int,
-    @DrawableRes val turnIconTurnSlightRight: Int,
-    @DrawableRes val turnIconTurnSharpLeft: Int,
-    @DrawableRes val turnIconTurnSharpRight: Int,
-    @DrawableRes val turnIconUturn: Int
-) {
+import com.mapbox.navigation.ui.base.model.TurnIconResources as BaseTurnIconResources
+
+class TurnIconResources private constructor(
```
Comment (abhishek1508) [inline libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/internal/TurnIconResources.kt:7]: I would revert the change to rename the file from `NotificationTurnIconResources` to `TurnIconResources`, just so that it is explicitly clear that these resources are to configure turn icons for notifications and there is no confusion with 2 files having the same name

File: libnavui-base/src/main/java/com/mapbox/navigation/ui/base/model/TurnIconResources.kt [hunk 3]
```diff
@@ -0,0 +1,267 @@
+package com.mapbox.navigation.ui.base.model
+
+/**
+ * An interface describing all maneuver icons.
+ */
+interface TurnIconResources {
```
Comment (abhishek1508) [inline libnavui-base/src/main/java/com/mapbox/navigation/ui/base/model/TurnIconResources.kt:6]: Can we name it as `BaseTurnIconResources` rather than using it as an alias in import statements?

File: libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/internal/TurnIconResources.kt (hunk 2 above)
Comment (tomaszrybakiewicz) [reply libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/internal/TurnIconResources.kt:7]: This class is a member of `.trip.notification` package. Adding `Notification` prefix is redundant, IMHO.

File: libnavui-base/src/main/java/com/mapbox/navigation/ui/base/model/TurnIconResources.kt (hunk 3 above)
Comment (tomaszrybakiewicz) [reply libnavui-base/src/main/java/com/mapbox/navigation/ui/base/model/TurnIconResources.kt:6]: What's wrong with `TurnIconResources` and using import aliases?

**Comment by tomaszrybakiewicz** [comment]
> We can see that we are not allowing external users to define their own set of TurnIconResources as we are defining it explicitly inside the class thereby losing the purpose of TurnIconResources. I think we should take this as a constructor parameter with default value being TurnIconResources.defaultIconSet() so it doesn't breaks SEMVER.

@abhishek1508 This change seems to be outside of the scope of this ticket. Would you like me to do this as a part of this PR?

File: libnavui-maneuver/src/test/java/com/mapbox/navigation/ui/maneuver/TurnIconHelperTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Guardiola31337) [reply libnavui-maneuver/src/test/java/com/mapbox/navigation/ui/maneuver/TurnIconHelperTest.kt:1348]: Ohh, why this test doesn't have `Arrange`? 🤔 

I'm not a big fan of Inline Setup https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=65

[the remark this answers is not in the document]
Comment (tomaszrybakiewicz) [reply libnavui-maneuver/src/test/java/com/mapbox/navigation/ui/maneuver/TurnIconHelperTest.kt:1348]: These tests are trivial enough that doesn't need `Arrange`.

**Comment by abhishek1508** [comment]
> > We can see that we are not allowing external users to define their own set of TurnIconResources as we are defining it explicitly inside the class thereby losing the purpose of TurnIconResources. I think we should take this as a constructor parameter with default value being TurnIconResources.defaultIconSet() so it doesn't breaks SEMVER.
> 
> @abhishek1508 This change seems to be outside of the scope of this ticket. Would you like me to do this as a part of this PR?

I would leave it up to you even though I think that it is probably a very small change that can happen in this PR itself since the fix does lines up with the `TurnIconResources` general fix you introduced in this PR.

File: libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/internal/TurnIconResources.kt (hunk 2 above)
Comment (abhishek1508) [reply libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/internal/TurnIconResources.kt:7]: In my opinion it improves the readability since now we have 3 `TurnIconResources` in the SDK and with multiple files opened the only way you can figure out which one belongs to which SDK is by looking at the package name which usually no one cares unless they need to import it.

File: libnavui-base/src/main/java/com/mapbox/navigation/ui/base/model/TurnIconResources.kt (hunk 3 above)
Comment (abhishek1508) [reply libnavui-base/src/main/java/com/mapbox/navigation/ui/base/model/TurnIconResources.kt:6]: There is nothing wrong with it, but again as pointed above, I think directly using the name in the original file improves the readability because import statements are usually collapsed and we don't tend to look at it. If someone were to find this file they won't find it because the alias is pretty much hidden in the collapse and this file actually exists by a different name. This happened with me in this PR.

**Comment by abhishek1508** [comment]
> > > > We can see that we are not allowing external users to define their own set of TurnIconResources as we are defining it explicitly inside the class thereby losing the purpose of TurnIconResources. I think we should take this as a constructor parameter with default value being TurnIconResources.defaultIconSet() so it doesn't breaks SEMVER.
> > > 
> > > @abhishek1508 This change seems to be outside of the scope of this ticket. Would you like me to do this as a part of this PR?
> > 
> > I would leave it up to you even though I think that it is probably a very small change that can happen in this PR itself since the fix does lines up with the `TurnIconResources` general fix you introduced in this PR. 
> 
> @abhishek1508 This unfortunately looks like a slightly bigger change. 
> 
> MapboxTripNotification is an internal class that is being instantiated via `MapboxModuleProvider.createModule` in MapboxNavigation. So if we want to allow overrides of default NotificationTurnIconResources icon set, we should probably update `MapboxNavigation.paramsProvider` and `NavigationOptions` as well.
> 
> ```kotlin
> private fun paramsProvider(type: MapboxModuleType): Array<ModuleProviderArgument> {
>         return when (type) {
>               ....
>               MapboxModuleType.NavigationTripNotification -> arrayOf(
>                   ModuleProviderArgument(NavigationOptions::class.java, navigationOptions),
>                   ModuleProviderArgument(
>                       DistanceFormatter::class.java,
>                       MapboxDistanceFormatter(navigationOptions.distanceFormatterOptions)
>                   ),
>                   ModuleProviderArgument(
>                      BaseTurnIconResources::class.java,
>                      navigationOptions.notificationIconResources   // <-- requires new NavigationOptions property
>                   )
>              )
>          }
> }
> ```
> The new `notificationIconResources` property would have to be added to `NavigationOptions` with default value of.... `NotificationTurnIconResources.defaultIconSet()`? `NotificationTurnIconResources` is an internal class in `libtrip-notification` module. It would have to be either moved to `'libnavigation-base` module or instantiated via reflection, just like `MapboxTripNotification`.
> 
> How would you like to proceed?
> 

Thanks for the detailed explanation. Let's cut a different ticket in this case. 

@tomaszrybakiewicz

**Comment by tomaszrybakiewicz** [comment]
@abhishek1508 I have applied all the changes that you suggested. Can you give it another look?

**Review by abhishek1508 — APPROVED** [review]
